File: TypeScript/Codewars/src/resolutions/Strings-Mix.ts
type IMix = {
	letter: string;
	quantity: number;
	local: string;
};

const Local = {
	S1: "1",
	S2: "2",
	Equal: "=",
};

export const mix = (s1: string, s2: string): string => {
	const output: IMix[] = [];

	const Regex = /[^a-z]/g;
	const arraySumLetters = new Set<string>([...s1.replace(Regex, "").split(""), ...s2.replace(Regex, "").split("")]);

	Array.from(arraySumLetters).map((letter) => {
		const response: IMix = { letter: "", local: "", quantity: 0 };

		const countS1 = countLetter(s1, letter);
		const countS2 = countLetter(s2, letter);

		if (countS1 > countS2) {
			response.quantity = countS1;
			response.local = Local.S1;
		} else if (countS2 > countS1) {
			response.quantity = countS2;
			response.local = Local.S2;
		} else {
			response.quantity = countS1;
			response.local = Local.Equal;
		}

		response.letter = letter;

		output.push(response);
	});

	return output
		.filter((value) => value.quantity > 1)
		.sort((a, b) => sortMix(a, b))
		.map((value) => `${value.local}:${value.letter.repeat(value.quantity)}`)
		.join("/");
};

function countLetter(arr: string, letter: string): number {
	return arr.split("").filter((value) => value === letter).length;
}

function sortMix(a: IMix, b: IMix): number {
	if (a.quantity !== b.quantity) return b.quantity - a.quantity;

	const order: Record<string, number> = { "1": 0, "2": 1, "=": 2 };
	if (order[a.local] !== order[b.local]) {
		return order[a.local] - order[b.local];
	}

	return a.letter.localeCompare(b.letter);
}

File: Csharp/CodeWars/src/Completed/StringsMix.cs
using System.Text.RegularExpressions;

namespace CodeWars.Completed;

public static class StringsMix
{
    private class MixItem
    {
        public char Letter { get; set; }
        public int Quantity { get; set; }
        public string Local { get; set; } = "";
    }

    private static readonly Regex NonLowercaseRegex = new Regex("[^a-z]", RegexOptions.Compiled);

    public static string Mix(string s1, string s2)
    {
        var output = new List<MixItem>();

        var letters1 = NonLowercaseRegex.Replace(s1 ?? string.Empty, string.Empty).ToCharArray();
        var letters2 = NonLowercaseRegex.Replace(s2 ?? string.Empty, string.Empty).ToCharArray();

        var letterSet = new HashSet<char>(letters1.Concat(letters2));

        foreach (var letter in letterSet.Where(ch => ch >= 'a' && ch <= 'z'))
        {
            var countS1 = CountLetter(s1, letter);
            var countS2 = CountLetter(s2, letter);

            var item = new MixItem
            {
                Letter = letter
            };

            if (countS1 > countS2)
            {
                item.Quantity = countS1;
                item.Local = "1";
            }
            else if (countS2 > countS1)
            {
                item.Quantity = countS2;
                item.Local = "2";
            }
            else
            {
                item.Quantity = countS1;
                item.Local = "=";
            }

            output.Add(item);
        }

        var result = output
            .Where(x => x.Quantity > 1)
            .OrderByDescending(x => x.Quantity)
            .ThenBy(x => LocalOrder(x.Local))
            .ThenBy(x => x.Letter)
            .Select(x => $"{x.Local}:{new string(x.Letter, x.Quantity)}");

        return string.Join("/", result);
    }

    private static int CountLetter(string? s, char letter)
    {
        if (string.IsNullOrEmpty(s)) return 0;
        int count = 0;
        foreach (var ch in s)
        {
            if (ch == letter) count++;
        }
        return count;
    }

    private static int LocalOrder(string local)
    {
        return local switch
        {
            "1" => 0,
            "2" => 1,
            "=" => 2,
            _ => 3
        };
    }
}